File: Project2/Baitoan9/bfs_multigraph.cpp
/**
 * @file bfs_multigraph.cpp
 * @brief Breadth-First Search implementation for Multi Graph
 * @details Supports graphs with parallel edges but NO self-loops using adjacency list representation
 */

#include <bits/stdc++.h>

using namespace std;

/**
 * @brief Represents a multi graph with BFS traversal capabilities
 */
class MultiGraph {
private:
    vector<vector<int>> adjacencyList;
    int numberOfVertices;
    vector<bool> visitedVertices;
    vector<int> traversalOrder;
    vector<int> distances;
    vector<int> parents;
    
    /**
     * @brief Resets visited status and other tracking arrays
     */
    void resetTraversalStatus() {
        fill(visitedVertices.begin(), visitedVertices.end(), false);
        traversalOrder.clear();
        fill(distances.begin(), distances.end(), -1);
        fill(parents.begin(), parents.end(), -1);
    }
    
    /**
     * @brief Validates that vertices are within valid range
     * @param vertex Vertex to validate
     * @return True if vertex is valid
     */
    bool isValidVertex(int vertex) const {
        return vertex >= 0 && vertex < numberOfVertices;
    }

public:
    /**
     * @brief Constructs a multi graph with specified number of vertices
     * @param vertexCount Number of vertices in the graph
     */
    explicit MultiGraph(int vertexCount) : numberOfVertices(vertexCount) {
        adjacencyList.resize(vertexCount);
        visitedVertices.resize(vertexCount, false);
        distances.resize(vertexCount, -1);
        parents.resize(vertexCount, -1);
    }
    
    /**
     * @brief Adds an edge to the multi graph with validation
     * @param sourceVertex Source vertex of the edge
     * @param targetVertex Target vertex of the edge
     * @return True if edge was added successfully
     */
    bool addEdge(int sourceVertex, int targetVertex) {
        if (!isValidVertex(sourceVertex) || !isValidVertex(targetVertex)) {
            cerr << "Warning: Invalid vertex index (" << sourceVertex 
                 << ", " << targetVertex << "). Vertices must be in range [0, " 
                 << numberOfVertices - 1 << "]\n";
            return false;
        }
        
        if (sourceVertex == targetVertex) {
            cerr << "Warning: Self-loop detected (" << sourceVertex 
                 << " -> " << targetVertex << "). Multi graphs do not support self-loops. Edge ignored.\n";
            return false;
        }
        
        // Multi graphs allow parallel edges but not self-loops
        adjacencyList[sourceVertex].push_back(targetVertex);
        adjacencyList[targetVertex].push_back(sourceVertex);
        return true;
    }
    
    /**
     * @brief Executes BFS traversal using queue-based approach
     * @param startVertex Starting vertex for traversal
     * @return Vector containing vertices in BFS order
     */
    vector<int> executeBFS(int startVertex) {
        if (!isValidVertex(startVertex)) {
            cerr << "Error: Invalid starting vertex " << startVertex << "\n";
            return vector<int>();
        }
        
        resetTraversalStatus();
        queue<int> bfsQueue;
        
        visitedVertices[startVertex] = true;
        distances[startVertex] = 0;
        bfsQueue.push(startVertex);
        
        while (!bfsQueue.empty()) {
            int currentVertex = bfsQueue.front();
            bfsQueue.pop();
            traversalOrder.push_back(currentVertex);
            
            for (int neighbor : adjacencyList[currentVertex]) {
                if (!visitedVertices[neighbor]) {
                    visitedVertices[neighbor] = true;
                    distances[neighbor] = distances[currentVertex] + 1;
                    parents[neighbor] = currentVertex;
                    bfsQueue.push(neighbor);
                }
            }
        }
        
        return traversalOrder;
    }
    
    /**
     * @brief Finds all connected components using BFS
     * @return Vector of vectors, each containing vertices of a connected component
     */
    vector<vector<int>> findConnectedComponents() {
        vector<vector<int>> components;
        resetTraversalStatus();
        
        for (int vertex = 0; vertex < numberOfVertices; ++vertex) {
            if (!visitedVertices[vertex]) {
                vector<int> component = executeBFSComponent(vertex);
                if (!component.empty()) {
                    components.push_back(component);
                }
            }
        }
        
        return components;
    }
    
    /**
     * @brief Gets the shortest distance to a vertex from last BFS start
     * @param vertex Target vertex
     * @return Distance to vertex, or -1 if unreachable
     */
    int getDistance(int vertex) const {
        if (!isValidVertex(vertex)) {
            return -1;
        }
        return distances[vertex];
    }
    
    /**
     * @brief Gets the parent of a vertex in BFS tree from last traversal
     * @param vertex Target vertex
     * @return Parent vertex, or -1 if no parent
     */
    int getParent(int vertex) const {
        if (!isValidVertex(vertex)) {
            return -1;
        }
        return parents[vertex];
    }
    
    /**
     * @brief Gets the shortest path from last BFS start to target vertex
     * @param targetVertex Target vertex
     * @return Vector containing path vertices, empty if no path exists
     */
    vector<int> getShortestPath(int targetVertex) const {
        if (!isValidVertex(targetVertex) || distances[targetVertex] == -1) {
            return vector<int>();
        }
        
        vector<int> path;
        int current = targetVertex;
        
        while (current != -1) {
            path.push_back(current);
            current = parents[current];
        }
        
        reverse(path.begin(), path.end());
        return path;
    }
    
    /**
     * @brief Gets the number of vertices in the graph
     * @return Number of vertices
     */
    int getVertexCount() const {
        return numberOfVertices;
    }
    
    /**
     * @brief Counts total number of edges including parallel edges
     * @return Total edge count
     */
    int getTotalEdgeCount() const {
        int totalEdges = 0;
        for (const auto& neighbors : adjacencyList) {
            totalEdges += static_cast<int>(neighbors.size());
        }
        return totalEdges / 2; // Each undirected edge counted twice (no self-loops in multi graph)
    }
    
    /**
     * @brief Displays the adjacency list representation with parallel edge indication
     */
    void displayGraph() const {
        cout << "\nMulti Graph - Adjacency List Representation:\n";
        cout << "Parallel edges allowed, NO self-loops\n";
        
        for (int vertex = 0; vertex < numberOfVertices; ++vertex) {
            cout << "Vertex " << vertex << ": ";
            if (!adjacencyList[vertex].empty()) {
                for (size_t i = 0; i < adjacencyList[vertex].size(); ++i) {
                    if (i > 0) cout << " -> ";
                    cout << adjacencyList[vertex][i];
                }
            }
            cout << "\n";
        }
        
        cout << "Total vertices: " << numberOfVertices << "\n";
        cout << "Total edges: " << getTotalEdgeCount() << "\n";
    }
    
    /**
     * @brief Displays parallel edge statistics
     */
    void displayParallelEdgeStatistics() const {
        map<pair<int, int>, int> edgeCount;
        
        for (int vertex = 0; vertex < numberOfVertices; ++vertex) {
            for (int neighbor : adjacencyList[vertex]) {
                if (vertex < neighbor) { // Count each edge only once
                    pair<int, int> edge = {vertex, neighbor};
                    edgeCount[edge]++;
                }
            }
        }
        
        cout << "\nParallel Edge Analysis:\n";
        bool hasParallelEdges = false;
        
        for (const auto& edgePair : edgeCount) {
            if (edgePair.second > 1) {
                hasParallelEdges = true;
                cout << "Edge (" << edgePair.first.first << ", " 
                     << edgePair.first.second << ") appears " 
                     << edgePair.second << " times\n";
            }
        }
        
        if (!hasParallelEdges) {
            cout << "No parallel edges detected in this multi graph.\n";
        }
    }
    
    /**
     * @brief Displays BFS tree information
     * @param startVertex Starting vertex of last BFS
     */
    void displayBFSTree(int startVertex) const {
        cout << "\nBFS Tree Information (from vertex " << startVertex << "):\n";
        cout << "Vertex | Distance | Parent\n";
        cout << "-------|----------|-------\n";
        
        for (int vertex = 0; vertex < numberOfVertices; ++vertex) {
            cout << setw(6) << vertex << " | ";
            if (distances[vertex] == -1) {
                cout << setw(8) << "INF" << " | ";
                cout << "N/A";
            } else {
                cout << setw(8) << distances[vertex] << " | ";
                if (parents[vertex] == -1) {
                    cout << "NIL";
                } else {
                    cout << parents[vertex];
                }
            }
            cout << "\n";
        }
    }

private:
    /**
     * @brief Helper function for finding connected components
     * @param startVertex Starting vertex for component search
     * @return Vector containing vertices in the component
     */
    vector<int> executeBFSComponent(int startVertex) {
        vector<int> component;
        queue<int> bfsQueue;
        
        visitedVertices[startVertex] = true;
        bfsQueue.push(startVertex);
        
        while (!bfsQueue.empty()) {
            int currentVertex = bfsQueue.front();
            bfsQueue.pop();
            component.push_back(currentVertex);
            
            for (int neighbor : adjacencyList[currentVertex]) {
                if (!visitedVertices[neighbor]) {
                    visitedVertices[neighbor] = true;
                    bfsQueue.push(neighbor);
                }
            }
        }
        
        return component;
    }
};

/**
 * @brief Handles input operations for multi graph construction
 */
class MultiGraphInputHandler {
private:
    istream& inputStream;
    
public:
    /**
     * @brief Constructs input handler with specified stream
     * @param stream Input stream to read from
     */
    explicit MultiGraphInputHandler(istream& stream) : inputStream(stream) {}
    
    /**
     * @brief Reads graph data and constructs MultiGraph
     * @return Constructed MultiGraph instance
     */
    unique_ptr<MultiGraph> readGraphData() {
        int vertexCount, edgeCount;
        inputStream >> vertexCount >> edgeCount;
        
        auto graph = make_unique<MultiGraph>(vertexCount);
        int successfulEdges = 0;
        
        for (int edgeIndex = 0; edgeIndex < edgeCount; ++edgeIndex) {
            int sourceVertex, targetVertex;
            inputStream >> sourceVertex >> targetVertex;
            if (graph->addEdge(sourceVertex, targetVertex)) {
                successfulEdges++;
            }
        }
        
        cout << "Successfully added " << successfulEdges 
             << " out of " << edgeCount << " edges to multi graph.\n";
        
        return graph;
    }
    
    /**
     * @brief Reads starting vertex for BFS traversal
     * @return Starting vertex index
     */
    int readStartingVertex() {
        int startVertex;
        inputStream >> startVertex;
        return startVertex;
    }
};

/**
 * @brief Handles output operations for BFS results
 */
class MultiGraphOutputHandler {
private:
    ostream& outputStream;
    
public:
    /**
     * @brief Constructs output handler with specified stream
     * @param stream Output stream to write to
     */
    explicit MultiGraphOutputHandler(ostream& stream) : outputStream(stream) {}
    
    /**
     * @brief Displays BFS traversal results
     * @param traversalResult Vector containing vertices in BFS order
     * @param startVertex Starting vertex
     */
    void displayTraversalResult(const vector<int>& traversalResult, int startVertex) {
        if (traversalResult.empty()) {
            outputStream << "BFS traversal: No traversal performed (invalid input)\n";
            return;
        }
        
        outputStream << "BFS traversal from vertex " << startVertex << ": ";
        for (size_t i = 0; i < traversalResult.size(); ++i) {
            if (i > 0) outputStream << " -> ";
            outputStream << traversalResult[i];
        }
        outputStream << "\n";
    }
    
    /**
     * @brief Displays connected components
     * @param components Vector of connected components
     */
    void displayConnectedComponents(const vector<vector<int>>& components) {
        outputStream << "\nConnected Components Analysis:\n";
        for (size_t i = 0; i < components.size(); ++i) {
            outputStream << "Component " << (i + 1) << ": ";
            for (size_t j = 0; j < components[i].size(); ++j) {
                if (j > 0) outputStream << " ";
                outputStream << components[i][j];
            }
            outputStream << "\n";
        }
        outputStream << "Total connected components: " << components.size() << "\n";
    }
    
    /**
     * @brief Displays shortest path information
     * @param graph Reference to the graph
     * @param startVertex Starting vertex
     * @param targetVertex Target vertex
     */
    void displayShortestPath(const MultiGraph& graph, int startVertex, int targetVertex) {
        vector<int> path = graph.getShortestPath(targetVertex);
        
        if (path.empty()) {
            outputStream << "No path exists from vertex " << startVertex 
                        << " to vertex " << targetVertex << "\n";
        } else {
            outputStream << "Shortest path from " << startVertex 
                        << " to " << targetVertex << ": ";
            for (size_t i = 0; i < path.size(); ++i) {
                if (i > 0) outputStream << " -> ";
                outputStream << path[i];
            }
            outputStream << " (distance: " << graph.getDistance(targetVertex) << ")\n";
        }
    }
    
    /**
     * @brief Displays program header information
     */
    void displayProgramHeader() {
        outputStream << "=== Breadth-First Search for Multi Graph ===\n";
        outputStream << "Supporting parallel edges, NO self-loops\n";
    }
};

/**
 * @brief Coordinates the Multi Graph BFS application workflow
 */
class MultiGraphBFSApplication {
private:
    unique_ptr<MultiGraph> graphInstance;
    unique_ptr<MultiGraphInputHandler> inputHandler;
    unique_ptr<MultiGraphOutputHandler> outputHandler;
    
public:
    /**
     * @brief Constructs BFS application with I/O streams
     * @param input Input stream for reading data
     * @param output Output stream for displaying results
     */
    MultiGraphBFSApplication(istream& input, ostream& output) {
        inputHandler = make_unique<MultiGraphInputHandler>(input);
        outputHandler = make_unique<MultiGraphOutputHandler>(output);
    }
    
    /**
     * @brief Executes the complete Multi Graph BFS application workflow
     */
    void executeApplication() {
        outputHandler->displayProgramHeader();
        graphInstance = inputHandler->readGraphData();
        int startingVertex = inputHandler->readStartingVertex();
        performBFSAnalysis(startingVertex);
    }

private:
    /**
     * @brief Performs comprehensive BFS analysis
     * @param startVertex Starting vertex for analysis
     */
    void performBFSAnalysis(int startVertex) {
        // Display graph structure
        graphInstance->displayGraph();
        graphInstance->displayParallelEdgeStatistics();
        
        // Perform BFS traversal
        vector<int> bfsResult = graphInstance->executeBFS(startVertex);
        outputHandler->displayTraversalResult(bfsResult, startVertex);
        
        // Display BFS tree information
        graphInstance->displayBFSTree(startVertex);
        
        // Find and display connected components
        vector<vector<int>> components = graphInstance->findConnectedComponents();
        outputHandler->displayConnectedComponents(components);
        
        // Display shortest path examples to all reachable vertices
        cout << "\nShortest Paths from vertex " << startVertex << ":\n";
        for (int target = 0; target < graphInstance->getVertexCount(); ++target) {
            if (target != startVertex && graphInstance->getDistance(target) != -1) {
                outputHandler->displayShortestPath(*graphInstance, startVertex, target);
            }
        }
    }
};

/**
 * @brief Main program entry point
 * @return Program exit status
 */
int main() {
    try {
        ifstream inputFile("input.txt");
        if (!inputFile.is_open()) {
            cerr << "Error: Cannot open input.txt file\n";
            return 1;
        }
        
        MultiGraphBFSApplication application(inputFile, cout);
        application.executeApplication();
        
        inputFile.close();
    } catch (const exception& error) {
        cerr << "Error occurred: " << error.what() << "\n";
        return 1;
    }
    return 0;
} 

File: Project2/Baitoan9/bfs_multigraph.py
#!/usr/bin/env python3
"""
@file bfs_multigraph.py
@brief Breadth-First Search implementation for Multi Graph
@details Supports graphs with parallel edges but NO self-loops using adjacency list representation
"""

from typing import List, Optional, Deque, Dict, Tuple
from collections import deque
import sys


class MultiGraph:
    """
    Represents a multi graph with BFS traversal capabilities
    """
    
    def __init__(self, vertex_count: int):
        """
        Constructs a multi graph with specified number of vertices
        
        Args:
            vertex_count: Number of vertices in the graph
        """
        self._adjacency_list: List[List[int]] = [[] for _ in range(vertex_count)]
        self._number_of_vertices: int = vertex_count
        self._visited_vertices: List[bool] = [False] * vertex_count
        self._traversal_order: List[int] = []
        self._distances: List[int] = [-1] * vertex_count
        self._parents: List[int] = [-1] * vertex_count
    
    def _reset_traversal_status(self) -> None:
        """
        Resets visited status and other tracking arrays
        """
        self._visited_vertices = [False] * self._number_of_vertices
        self._traversal_order.clear()
        self._distances = [-1] * self._number_of_vertices
        self._parents = [-1] * self._number_of_vertices
    
    def _is_valid_vertex(self, vertex: int) -> bool:
        """
        Validates that vertices are within valid range
        
        Args:
            vertex: Vertex to validate
            
        Returns:
            True if vertex is valid
        """
        return 0 <= vertex < self._number_of_vertices
    
    def add_edge(self, source_vertex: int, target_vertex: int) -> bool:
        """
        Adds an edge to the multi graph with validation
        
        Args:
            source_vertex: Source vertex of the edge
            target_vertex: Target vertex of the edge
            
        Returns:
            True if edge was added successfully
        """
        if not self._is_valid_vertex(source_vertex) or not self._is_valid_vertex(target_vertex):
            print(f"Warning: Invalid vertex index ({source_vertex}, {target_vertex}). "
                  f"Vertices must be in range [0, {self._number_of_vertices - 1}]", file=sys.stderr)
            return False
        
        if source_vertex == target_vertex:
            print(f"Warning: Self-loop detected ({source_vertex} -> {target_vertex}). "
                  f"Multi graphs do not support self-loops. Edge ignored.", file=sys.stderr)
            return False
        
        # Multi graphs allow parallel edges but not self-loops
        self._adjacency_list[source_vertex].append(target_vertex)
        self._adjacency_list[target_vertex].append(source_vertex)
        return True
    
    def execute_bfs(self, start_vertex: int) -> List[int]:
        """
        Executes BFS traversal using queue-based approach
        
        Args:
            start_vertex: Starting vertex for traversal
            
        Returns:
            List containing vertices in BFS order
        """
        if not self._is_valid_vertex(start_vertex):
            print(f"Error: Invalid starting vertex {start_vertex}", file=sys.stderr)
            return []
        
        self._reset_traversal_status()
        bfs_queue: Deque[int] = deque()
        
        self._visited_vertices[start_vertex] = True
        self._distances[start_vertex] = 0
        bfs_queue.append(start_vertex)
        
        while bfs_queue:
            current_vertex = bfs_queue.popleft()
            self._traversal_order.append(current_vertex)
            
            for neighbor in self._adjacency_list[current_vertex]:
                if not self._visited_vertices[neighbor]:
                    self._visited_vertices[neighbor] = True
                    self._distances[neighbor] = self._distances[current_vertex] + 1
                    self._parents[neighbor] = current_vertex
                    bfs_queue.append(neighbor)
        
        return self._traversal_order.copy()
    
    def find_connected_components(self) -> List[List[int]]:
        """
        Finds all connected components using BFS
        
        Returns:
            List of lists, each containing vertices of a connected component
        """
        components: List[List[int]] = []
        self._reset_traversal_status()
        
        for vertex in range(self._number_of_vertices):
            if not self._visited_vertices[vertex]:
                component = self._execute_bfs_component(vertex)
                if component:
                    components.append(component)
        
        return components
    
    def get_distance(self, vertex: int) -> int:
        """
        Gets the shortest distance to a vertex from last BFS start
        
        Args:
            vertex: Target vertex
            
        Returns:
            Distance to vertex, or -1 if unreachable
        """
        if not self._is_valid_vertex(vertex):
            return -1
        return self._distances[vertex]
    
    def get_parent(self, vertex: int) -> int:
        """
        Gets the parent of a vertex in BFS tree from last traversal
        
        Args:
            vertex: Target vertex
            
        Returns:
            Parent vertex, or -1 if no parent
        """
        if not self._is_valid_vertex(vertex):
            return -1
        return self._parents[vertex]
    
    def get_shortest_path(self, target_vertex: int) -> List[int]:
        """
        Gets the shortest path from last BFS start to target vertex
        
        Args:
            target_vertex: Target vertex
            
        Returns:
            List containing path vertices, empty if no path exists
        """
        if not self._is_valid_vertex(target_vertex) or self._distances[target_vertex] == -1:
            return []
        
        path: List[int] = []
        current = target_vertex
        
        while current != -1:
            path.append(current)
            current = self._parents[current]
        
        path.reverse()
        return path
    
    def get_vertex_count(self) -> int:
        """
        Gets the number of vertices in the graph
        
        Returns:
            Number of vertices
        """
        return self._number_of_vertices
    
    def get_total_edge_count(self) -> int:
        """
        Counts total number of edges including parallel edges
        
        Returns:
            Total edge count
        """
        total_edges = sum(len(neighbors) for neighbors in self._adjacency_list)
        return total_edges // 2  # Each undirected edge counted twice (no self-loops in multi graph)
    
    def display_graph(self) -> str:
        """
        Returns the adjacency list representation with parallel edge indication
        
        Returns:
            String representation of adjacency list
        """
        result = "\nMulti Graph - Adjacency List Representation:\n"
        result += "Parallel edges allowed, NO self-loops\n"
        
        for vertex in range(self._number_of_vertices):
            result += f"Vertex {vertex}: "
            if self._adjacency_list[vertex]:
                result += " -> ".join(map(str, self._adjacency_list[vertex]))
            result += "\n"
        
        result += f"Total vertices: {self._number_of_vertices}\n"
        result += f"Total edges: {self.get_total_edge_count()}\n"
        
        return result
    
    def display_parallel_edge_statistics(self) -> str:
        """
        Returns parallel edge statistics
        
        Returns:
            String with parallel edge analysis
        """
        edge_count: Dict[Tuple[int, int], int] = {}
        
        for vertex in range(self._number_of_vertices):
            for neighbor in self._adjacency_list[vertex]:
                if vertex < neighbor:  # Count each edge only once
                    edge = (vertex, neighbor)
                    edge_count[edge] = edge_count.get(edge, 0) + 1
        
        result = "\nParallel Edge Analysis:\n"
        has_parallel_edges = False
        
        for edge, count in edge_count.items():
            if count > 1:
                has_parallel_edges = True
                result += f"Edge ({edge[0]}, {edge[1]}) appears {count} times\n"
        
        if not has_parallel_edges:
            result += "No parallel edges detected in this multi graph.\n"
        
        return result
    
    def display_bfs_tree(self, start_vertex: int) -> str:
        """
        Returns BFS tree information
        
        Args:
            start_vertex: Starting vertex of last BFS
            
        Returns:
            String with BFS tree information
        """
        result = f"\nBFS Tree Information (from vertex {start_vertex}):\n"
        result += "Vertex | Distance | Parent\n"
        result += "-------|----------|-------\n"
        
        for vertex in range(self._number_of_vertices):
            result += f"{vertex:6} | "
            if self._distances[vertex] == -1:
                result += f"{'INF':8} | N/A"
            else:
                result += f"{self._distances[vertex]:8} | "
                if self._parents[vertex] == -1:
                    result += "NIL"
                else:
                    result += str(self._parents[vertex])
            result += "\n"
        
        return result
    
    def _execute_bfs_component(self, start_vertex: int) -> List[int]:
        """
        Helper function for finding connected components
        
        Args:
            start_vertex: Starting vertex for component search
            
        Returns:
            List containing vertices in the component
        """
        component: List[int] = []
        bfs_queue: Deque[int] = deque()
        
        self._visited_vertices[start_vertex] = True
        bfs_queue.append(start_vertex)
        
        while bfs_queue:
            current_vertex = bfs_queue.popleft()
            component.append(current_vertex)
            
            for neighbor in self._adjacency_list[current_vertex]:
                if not self._visited_vertices[neighbor]:
                    self._visited_vertices[neighbor] = True
                    bfs_queue.append(neighbor)
        
        return component


class MultiGraphInputHandler:
    """
    Handles input operations for multi graph construction
    """
    
    def __init__(self, input_source):
        """
        Constructs input handler with specified source
        
        Args:
            input_source: Input source to read from (file object or similar)
        """
        self._input_source = input_source
    
    def read_graph_data(self) -> MultiGraph:
        """
        Reads graph data and constructs MultiGraph
        
        Returns:
            Constructed MultiGraph instance
        """
        line = self._input_source.readline().strip()
        vertex_count, edge_count = map(int, line.split())
        
        graph = MultiGraph(vertex_count)
        successful_edges = 0
        
        for _ in range(edge_count):
            line = self._input_source.readline().strip()
            source_vertex, target_vertex = map(int, line.split())
            if graph.add_edge(source_vertex, target_vertex):
                successful_edges += 1
        
        print(f"Successfully added {successful_edges} out of {edge_count} edges to multi graph.")
        
        return graph
    
    def read_starting_vertex(self) -> int:
        """
        Reads starting vertex for BFS traversal
        
        Returns:
            Starting vertex index
        """
        line = self._input_source.readline().strip()
        return int(line)


class MultiGraphOutputHandler:
    """
    Handles output operations for BFS results
    """
    
    def __init__(self, output_target=sys.stdout):
        """
        Constructs output handler with specified target
        
        Args:
            output_target: Output target to write to
        """
        self._output_target = output_target
    
    def display_traversal_result(self, traversal_result: List[int], start_vertex: int) -> None:
        """
        Displays BFS traversal results
        
        Args:
            traversal_result: List containing vertices in BFS order
            start_vertex: Starting vertex
        """
        if not traversal_result:
            self._output_target.write("BFS traversal: No traversal performed (invalid input)\n")
            return
        
        result_str = " -> ".join(map(str, traversal_result))
        self._output_target.write(f"BFS traversal from vertex {start_vertex}: {result_str}\n")
    
    def display_connected_components(self, components: List[List[int]]) -> None:
        """
        Displays connected components
        
        Args:
            components: List of connected components
        """
        self._output_target.write("\nConnected Components Analysis:\n")
        for i, component in enumerate(components, 1):
            component_str = " ".join(map(str, component))
            self._output_target.write(f"Component {i}: {component_str}\n")
        self._output_target.write(f"Total connected components: {len(components)}\n")
    
    def display_shortest_path(self, graph: MultiGraph, start_vertex: int, target_vertex: int) -> None:
        """
        Displays shortest path information
        
        Args:
            graph: Reference to the graph
            start_vertex: Starting vertex
            target_vertex: Target vertex
        """
        path = graph.get_shortest_path(target_vertex)
        
        if not path:
            self._output_target.write(f"No path exists from vertex {start_vertex} to vertex {target_vertex}\n")
            else:
            path_str = " -> ".join(map(str, path))
            distance = graph.get_distance(target_vertex)
            self._output_target.write(f"Shortest path from {start_vertex} to {target_vertex}: {path_str} (distance: {distance})\n")
    
    def display_program_header(self) -> None:
        """
        Displays program header information
        """
        self._output_target.write("=== Breadth-First Search for Multi Graph ===\n")
        self._output_target.write("Supporting parallel edges, NO self-loops\n")


class MultiGraphBFSApplication:
    """
    Coordinates the Multi Graph BFS application workflow
    """
    
    def __init__(self, input_source, output_target=sys.stdout):
        """
        Constructs BFS application with I/O sources
        
        Args:
            input_source: Input source for reading data
            output_target: Output target for displaying results
        """
        self._graph_instance: Optional[MultiGraph] = None
        self._input_handler = MultiGraphInputHandler(input_source)
        self._output_handler = MultiGraphOutputHandler(output_target)
    
    def execute_application(self) -> None:
        """
        Executes the complete Multi Graph BFS application workflow
        """
        self._output_handler.display_program_header()
        self._graph_instance = self._input_handler.read_graph_data()
        starting_vertex = self._input_handler.read_starting_vertex()
        self._perform_bfs_analysis(starting_vertex)
    
    def _perform_bfs_analysis(self, start_vertex: int) -> None:
        """
        Performs comprehensive BFS analysis
        
        Args:
            start_vertex: Starting vertex for analysis
        """
        if self._graph_instance is None:
            raise ValueError("Graph instance not initialized")
        
        # Display graph structure
        print(self._graph_instance.display_graph())
        print(self._graph_instance.display_parallel_edge_statistics())
        
        # Perform BFS traversal
        bfs_result = self._graph_instance.execute_bfs(start_vertex)
        self._output_handler.display_traversal_result(bfs_result, start_vertex)
        
        # Display BFS tree information
        print(self._graph_instance.display_bfs_tree(start_vertex))
        
        # Find and display connected components
        components = self._graph_instance.find_connected_components()
        self._output_handler.display_connected_components(components)
        
        # Display shortest path examples to all reachable vertices
        print(f"\nShortest Paths from vertex {start_vertex}:")
        for target in range(self._graph_instance.get_vertex_count()):
            if target != start_vertex and self._graph_instance.get_distance(target) != -1:
                self._output_handler.display_shortest_path(self._graph_instance, start_vertex, target)


def main() -> int:
    """
    Main program entry point
    
    Returns:
        Program exit status
    """
    try:
        with open("input.txt", "r", encoding="utf-8") as input_file:
            application = MultiGraphBFSApplication(input_file)
            application.execute_application()
    except FileNotFoundError:
        print("Error: Cannot open input.txt file", file=sys.stderr)
        return 1
    except Exception as error:
        print(f"Error occurred: {error}", file=sys.stderr)
        return 1
    
    return 0


if __name__ == "__main__":
    sys.exit(main()) 